credentials/alts/internal/conn/record.go

```diff
@@ -120,16 +121,19 @@ func NewConn(c net.Conn, side core.Side, recordProtocol string, key []byte, prot
 		// altsRecordDefaultLength (bytes) data into protected at one
 		// time. Therefore, 2*altsRecordDefaultLength-1 is large enough
 		// to buffer data read from the network.
-		protected = make([]byte, 0, 2*altsRecordDefaultLength-1)
+		protectedBuf = make([]byte, 0, 2*altsRecordDefaultLength-1)
+	} else {
+		protectedBuf = make([]byte, len(protected))
```

Comment:
As I mentioned offline,  the buf in use before this change was 64k, but now is much smaller. So it's possible that protectedBuf doesn't have enough space for the next Read().

The approach we are doing now is to try to keep the previous behavior to be safe. But with a second thought, this isn't safe, because the buf is shrinked.

What symptoms will we see if the buf is too small? Will the messages be broken?

Reply:
`Read` (below) keeps reading until a complete frame is received. If the read buffer is exhausted its size will be increased by `altsRecordDefaultLength`. This keeps repeating until a full frame is received.

We can change the code in this PR to allocate at least `2*altsRecordDefaultLength-1`. Basically, if `len(protected) < 2*altsRecordDefaultLength-1`, we allocate `2*altsRecordDefaultLength-1` and copy. If not, we allocate `len(protected)` and copy. WDYT?

Reply:
Allocating `max(len(protected), 2*altsRecordDefaultLength-1)` sounds good.

There's still potential risk that `len(protected)` is big, and conn ends up keeping a quite big buf around. But I think we can start with this.

Reply:
I changed the code allocate `max(len(protected), 2*altsRecordDefaultLength-1)`.

However, I don't think we have the potential risk you mentioned. Initially, protected might contain encrypted application data that was read in the same network `read` with the last message of the handshake. Assume this data is larger than `2*altsRecordDefaultLength-1`, then we allocate `len(protected)`. Now assume that the application data was not read in the same network `read` with the last handshake message, in this case, we only allocate `2*altsRecordDefaultLength-1` because `protected` has 0 length initially. However, when the record protocol attempts to read the next frame, which is the application data that is larger than `2*altsRecordDefaultLength-1`, it will have to expand the allocated buffer to fit the frame. In both cases, we end up with a buffer of size greater than `2*altsRecordDefaultLength-1`, we either allocate it in `conn.NewConn` or the next time `Read` is called.